### PR TITLE
Task 4: cart update,delete cart item, MiniCart, selfcheck(add2cart in PLP)

### PR DESCRIPTION
**Pull request by egorzhukovskij**
Complete task 4.1, 4.2, 4.3, 4.4, 4.5[selfcheck]

### CODE REVIEW

File: src/main/java/com/es/phoneshop/web/CartPageServlet.java [hunk 1]
```diff
@@ -9,23 +10,70 @@
...
+        }
+    }
+
+    private int parseQuantity(String quantityString, HttpServletRequest request) throws ParseException {
+        NumberFormat format = NumberFormat.getInstance(request.getLocale());
+        return format.parse(quantityString).intValue();
+    }
+
+    private void getMessageException(Map<String, String> errors, String productCode, Exception e) {
+        if (e.getClass() == ParseException.class) {
```
Comment (yan-piatkevich) [inline src/main/java/com/es/phoneshop/web/CartPageServlet.java:70]: Why not instanceof?

File: src/main/java/com/es/phoneshop/web/CartPageServlet.java [hunk 2]
```diff
@@ -9,23 +10,70 @@
...
+            errors.put(productCode, "Not a number!");
+        } else if (e.getClass() == OutOfStockException.class) {
+            OutOfStockException ofStockException = (OutOfStockException) e;
```
Comment (yan-piatkevich) [inline src/main/java/com/es/phoneshop/web/CartPageServlet.java]: outOfStockException

File: src/main/java/com/es/phoneshop/web/MiniCartServlet.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    private static final String CART_ATTRIBUTE = "cart";
+    private static final String MINI_CART_JSP_PATH = "/WEB-INF/pages/miniCart.jsp";
+
+    private CartService cartService;
+
+    @Override
+    public void init(ServletConfig config) throws ServletException {
+        super.init(config);
+        cartService = CartServiceImpl.getInstance();
+
```
Comment (yan-piatkevich) [inline src/main/java/com/es/phoneshop/web/MiniCartServlet.java]: redundant line

File: src/main/java/com/es/phoneshop/web/ProductListPageServlet.java [hunk 4]
```diff
@@ -45,8 +56,48 @@ protected void doGet(HttpServletRequest request, HttpServletResponse response) t
...
+    }
+
+    private int parseQuantity(String quantityString, HttpServletRequest request) throws ParseException {
+        NumberFormat format = NumberFormat.getInstance(request.getLocale());
+        return format.parse(quantityString).intValue();
+    }
+
+
+    private String getMessageException(Exception e) {
+        if (e.getClass() == ParseException.class) {
```
Comment (yan-piatkevich) [inline src/main/java/com/es/phoneshop/web/ProductListPageServlet.java:94]: Why not instanceof?

File: src/test/java/com/es/phoneshop/web/ProductDetailsPageServletTest.java [hunk 5]
```diff
@@ -108,9 +127,22 @@ public void givenRequestResponse_WhenDoPost_ThenInvokedSendRedirectOneTime() thr
...
+
+    @Test
+    public void testDoPostCartServiceInvokedAdd() throws OutOfStockException, IOException, ServletException {
+        when(cartService.getCart(any())).thenReturn(cart);
+        when(request.getLocale()).thenReturn(Locale.ENGLISH);
+        when(request.getParameter(QUANTITY_PARAM)).thenReturn("2");
+
+        servlet.doPost(request, response);
+
+        verify(cartService, atLeast(1)).add(cart, MOCKED_PATH_INFO.substring(1), 2);
```
Comment (yan-piatkevich) [inline src/test/java/com/es/phoneshop/web/ProductDetailsPageServletTest.java]: 2 is magic number

File: src/test/java/com/es/phoneshop/web/ProductListPageServletTest.java [hunk 6]
```diff
@@ -43,7 +77,61 @@ public void testDoGet() throws ServletException, IOException {
...
+    public void testDоPostSendRedirectIfQueryNotEmpty() throws ServletException, IOException, OutOfStockException {
+        when(cartService.getCart(any())).thenReturn(cart);
+        when(request.getLocale()).thenReturn(Locale.ENGLISH);
+        when(request.getParameter(PRODUCT_CODE_PARAM)).thenReturn("");
+        when(request.getParameter(QUANTITY_PARAM)).thenReturn("2");
+        when(request.getParameter(QUERY_PARAM)).thenReturn("iphone");
+
+        servlet.doPost(request, response);
+
+        verify(cartService, atLeast(1)).add(cart, "", 2);
```
Comment (yan-piatkevich) [inline src/test/java/com/es/phoneshop/web/ProductListPageServletTest.java]: 2 is magic number

File: src/main/java/com/es/phoneshop/service/impl/CartServiceImpl.java [hunk 7]
```diff
@@ -47,22 +50,16 @@ public Cart getCart(HttpSession session) {
...
         synchronized (LOCK) {
-            if (cart == null) {
-                throw new NullPointerException("Cart is nulL");
-            }
             Product product = productDao.getProduct(productCode);
             if (quantity <= 0) {
-                throw new IllegalArgumentException("Quantity must be more than 0");
+                throw new IllegalArgumentException("Quantity must be more Then 0");
             } else if (product.getStock() < quantity) {
                 throw new OutOfStockException(product, quantity, product.getStock());
```
Comment (yan-piatkevich) [inline src/main/java/com/es/phoneshop/service/impl/CartServiceImpl.java]: Make OutOfStockException effectively immutable

File: src/main/java/com/es/phoneshop/web/CartPageServlet.java (hunk 1 above)
Comment (egorzhukovskij) [reply src/main/java/com/es/phoneshop/web/CartPageServlet.java:70]: **getClass** is faster than **instanceof**. If I am not mistaken, then **getClass** does not look at the inheritance hierarchy and will be more specific.

File: src/main/java/com/es/phoneshop/web/ProductListPageServlet.java (hunk 4 above)
Comment (egorzhukovskij) [reply src/main/java/com/es/phoneshop/web/ProductListPageServlet.java:94]: **getClass** is faster than **instanceof**. If I am not mistaken, then **getClass** does not look at the inheritance hierarchy and will be more specific.

**Review by yan-piatkevich — APPROVED** [review]

**Pull request merged**